### PR TITLE
Optimized curl transform formulas

### PR DESCRIPTION
**Pull request by muXxer**
powsrv.io Team optimized the curl transform formulas.

On my machine it's around 7% faster than before.

# Test Plan:
To benchmark the results I changed the test_curlp_ptrit.c locally on my machine:
```
int main(void) {
  UNITY_BEGIN();

  RUN_TEST(test_curl_p_27_works);
  for(size_t i = 0; i < 50000; i++)
  {
    RUN_TEST(test_curl_p_81_works);
  } 

  return UNITY_END();
}
```

### CODE REVIEW

**Comment by thibault-martinez** [comment]
Some benchmarks done in https://github.com/iotaledger/iri/pull/1330

File: common/curl-p/ptrit.c [hunk 1]
```diff
@@ -9,17 +9,16 @@
...
     alpha = s[CURL_INDEX[i]].low;
     beta = s[CURL_INDEX[i]].high;
-    gamma = s[CURL_INDEX[i + 1]].high;
-    delta = (alpha | (~gamma)) & (s[CURL_INDEX[i + 1]].low ^ beta);
+    delta = beta ^ s[CURL_INDEX[i + 1]].low;
 
-    c[i].low = ~delta;
-    c[i].high = (alpha ^ gamma) | delta;
+    c[i].low = ~(delta & alpha);
+    c[i].high = delta | (alpha ^ s[CURL_INDEX[i + 1]].high);
```
Comment (Wollac) [inline common/curl-p/ptrit.c]:
```suggestion
    c[i].high = (alpha ^ s[CURL_INDEX[i + 1]].high) | delta;
```
The formula for `c[i].high` remains unchanged.

File: common/curl-p/ptrit.c [hunk 2]
```diff
@@ -9,17 +9,16 @@
...
-  ptrit_s alpha, beta, gamma, delta;
+  ptrit_s alpha, beta, delta;
   size_t i = 0;
 
   for (; i < STATE_LENGTH; ++i) {
     alpha = s[CURL_INDEX[i]].low;
     beta = s[CURL_INDEX[i]].high;
-    gamma = s[CURL_INDEX[i + 1]].high;
-    delta = (alpha | (~gamma)) & (s[CURL_INDEX[i + 1]].low ^ beta);
+    delta = beta ^ s[CURL_INDEX[i + 1]].low;
```
Comment (Wollac) [inline common/curl-p/ptrit.c]:
```suggestion
    delta = alpha & (s[CURL_INDEX[i + 1]].low ^ beta);
```

File: common/curl-p/ptrit.c [hunk 3]
```diff
@@ -9,17 +9,16 @@
...
 
-    c[i].low = ~delta;
-    c[i].high = (alpha ^ gamma) | delta;
+    c[i].low = ~(delta & alpha);
```
Comment (Wollac) [inline common/curl-p/ptrit.c]:
```suggestion
    c[i].low = ~delta;
```

**Review by Wollac — COMMENTED** [review]
The changes reduce the number of bitwise operations, while still being correct under the assumption that `.low` and `.high` never *both* have a 0-bit at the same index, which is exactly a characteristic of the `ptrit` encoding.
This assumption is correct, if e.g. `trits_to_ptrits` is used for the conversion and it is a loop invariant in `ptrit_sbox`, but it is never explicitly validated before entering this function.

I proposed a few changes to the order, to make it clearer which parts of the computation have actually been changed.
These changes can be validated more formally for every bit:

review

File: common/curl-p/ptrit.c (hunk 1 above)
Comment (tsvisabo) [reply common/curl-p/ptrit.c]: Hi @muXxer , thanks for this PR, can you check if the performance gain is still the same after @Wollac  order changes suggestions and integrate them to your PR?

**Review by tsvisabo — CHANGES_REQUESTED** [review]

**Comment by muXxer** [comment]
Performance gain is the same with wollacs changes.

**Review by tsvisabo — DISMISSED** [review]

**Comment by tsvisabo** [comment]
@muXxer can you rebase and i'll re-approve?

**Comment by muXxer** [comment]
@tsvisabo done :)

**Review by tsvisabo — APPROVED** [review]

**Pull request merged**